ctlcmd/cmdbase.py

```diff
@@ -771,7 +772,198 @@ def globcomp(text):
+  measured SiPM temp (C), and measured pulser board temp (C) will always be in the root file. 
+  Extra branches with more data will be added based on the commands run. 
+  
+  The save file string can also include placeholder strings to have the filename
+  automatically be modified according other input arguments and/or the state of
+  the system. This allows the same command to be used for different
+  configurations while still have the results be saved in distinct files.
+  Placeholders will be specified in the angle braced. For the full list of
+  special placeholders, as well as how the file parsing is handled, see the
+  detailed documentation in the `ctlcmd.cmdbase.savefilecmd.parse` method.
+  """
+  ##TODO: update this to make the file location customizable
+  DEFAULT_ROOTFILE = 'SAVEROOT_<TIMESTAMP>'
+  def __init__(self, cmd):
+    print("initialize rootfilecmd")
+    controlcmd.__init__(self, cmd)
+    
+  def add_args(self):
+    print("add_args in rootfilecmd")
+    group = self.parser.add_argument_group(
+        "root file saving options", """Options for changing the root file location.
+        For more details, see the official documentation.""")
+    group.add_argument('--saveroot',
+                       type=str,
+                       default=self.DEFAULT_ROOTFILE,
+                       help="""File path to save (placeholders in angle braces).
+                       default=%(default)s""")
+ 
+  def parse(self,args):
+    print("parse in rootfilecmd")
+    if not args.saveroot:  # Early exit if savefile is not set
+      self.saveroot = None
+      return args
+    filename = args.saveroot
+
+    # Adding time stamp filenames
+    timestring = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+    filename = re.sub('<TIMESTAMP>', timestring, filename, flags=re.IGNORECASE)
+
+    # Adding boardid to the settings
+    filename = re.sub('<BOARDID>',
+                      str(self.board.boardid),
+                      filename,
+                      flags=re.IGNORECASE)
+    # Adding boardtype to the settings
+    filename = re.sub('<BOARDTYPE>',
+                      str(self.board.boardtype),
+                      filename,
+                      flags=re.IGNORECASE)
 
+    # Substituting tokens for argument values
+    for action in self.parser._actions:
+      for string in action.option_strings:
+        string = string.strip('-')
+        if not hasattr(args, string): continue
+        val = ''
+        try:
+          val = action.type(getattr(args, string))
+        except:
+          pass
+        substring = '{0}{1}'.format(string, val)
+        substring = re.sub(r'\.', 'p', substring)
+        filename = re.sub('\\<{0}\\>'.format(string),
+                          substring,
+                          filename,
+                          flags=re.IGNORECASE)
+      self.saveroot=filename
+      return args
+      
+  def openroot(self,functiontype):
```

Comment:
I am confused about this comment, doesn't the string filename need to be saved after it has been parsed so that the file can be set with that name? The root file location and name has not been set yet at this point.

Reply:
Ah what I meant is that you don't need to add a `functiontype` as an argument. You can simply call `self.__class__.__name__`. Inherited classes will have the class name automatically determined, even in base class methods.